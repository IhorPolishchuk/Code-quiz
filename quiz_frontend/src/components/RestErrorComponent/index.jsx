import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import {
    HOME_PATH
} from './../../routes';

import styles from './index.sass';

/**
 * Component that represent user rest error 
 * 
 * Default error is 404, but it can be 500 or any other, you can specify redirect to any path, code and message of error.
 * 
 * @prop {number} code          : number of error happend, default is 404
 * @prop {string} message       : message of error, deafult is message that you get wrong route 
 * @prop {bool} redirect        : if true enables button for redirect
 * @prop {string} redirectPath  : if redirect true, clicking on button you can redirect to givven path, default is home path
 * 
 * @class RestErrorComponent
 * @extends {Component}
 */
class RestErrorComponent extends Component {
    render () {
        const { code, message, redirect, redirectPath } = this.props;
        return (
            <div
                className={styles.nmc__wrapper}
            >
                <div 
                    className={styles.nmc__content}
                >
                    <h2
                        className={styles.nmc__error_code}
                    >
                        {code}
                    </h2>
                    <div>
                        <div
                            className={styles.nmc__text}
                        >
                            <div>{message}</div>
                            {redirect &&
                                <div>
                                    <div>You feel lost?</div>
                                    <Link to={redirectPath} className={styles.button__back}>Back Home</Link>
                                </div>
                            }
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

RestErrorComponent.propTypes = {
    code: PropTypes.number,
    message: PropTypes.string,
    redirect: PropTypes.bool,
    redirectPath: PropTypes.string
};  

RestErrorComponent.defaultProps = {
    code: 404,
    message: `Page was not found`,
    redirect: true,
    redirectPath: HOME_PATH
};

export default RestErrorComponent;
