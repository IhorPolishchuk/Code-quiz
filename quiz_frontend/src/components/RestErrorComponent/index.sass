$nmc_bg_color: #ffffff
$button_color: #ffffff
$button_bg: #9680aa
.nmc
    &__wrapper
        display: flex
        flex-direction: row
        justify-content: center
        position: absolute
        left: 0
        top: 0
        width: 100%
        height: 100vh
        background: $nmc_bg_color
    &__content
        margin-top: 200px
    &__error_code
        font-family: RobotoBold
        font-size: 48px
        font-style: italic
        text-align: center
    &__text
        display: flex
        flex-direction: column
        align-items: center
        font-family: RobotoRegular
        line-height: 1.5

.button__back
    margin: 0px 5px
    padding: 5px
    font-family: RobotoMedium
    color: $button_color
    background-color: $button_bg
    border-radius: 10px
    border-color: lighter($button_bg ,10%)