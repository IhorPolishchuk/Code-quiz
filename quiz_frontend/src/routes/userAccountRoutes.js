import * as routes from './index.js';

export default [
    {
        route: routes.USER_ACCOUNT_PATH,
        label: 'Profile',
        icon: 'user'
    },
    {
        route: routes.USER_SETTINGS_PATH,
        label: 'Settings',
        icon: 'setting'
    },
    {
        route: routes.USER_STATISTICS_PATH,
        label: 'Statistics',
        icon: 'area-chart'
    }
];
