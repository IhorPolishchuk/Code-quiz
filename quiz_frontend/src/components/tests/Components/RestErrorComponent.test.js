import React from 'react';
import { Link } from 'react-router-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import RestErrorComponent from './../../RestErrorComponent';

import { HOME_PATH } from './../../../routes';

const expectedDefaultProps = {
    code: 404,
    message: 'Page was not found',
    redirect: true,
    redirectPath: HOME_PATH
};
const expectedCustomProps = {
    code: 500,
    message: 'server error',
    redirect: false
};

describe('RestErrorComponent', () => {
    describe('RestErrorComponent with default props', () => {
        const component = shallow(
            <RestErrorComponent/>
        );

        it('Renders with expected props', () => {
            expect(component.instance().props.code).toBe(expectedDefaultProps.code);
            expect(component.instance().props.message).toBe(expectedDefaultProps.message);
            expect(component.instance().props.redirect).toBe(expectedDefaultProps.redirect);
            expect(component.instance().props.redirectPath).toBe(expectedDefaultProps.redirectPath);
        });

        it('Renders redirect link', () => {
            expect(component.find(Link)).toHaveLength(1);
        });

        it('Matching snapshot' , () => {
            const tree = toJson(component); 
            
            expect(tree).toMatchSnapshot();
        });
    });
    describe('RestErrorComponent with custom props', () => {
        const component = shallow(
            <RestErrorComponent
                code = {expectedCustomProps.code}
                message = {expectedCustomProps.message}
                redirect = {expectedCustomProps.redirect}
            />
        );

        it('Renders with expected props', () => {
            expect(component.instance().props.code).toBe(expectedCustomProps.code);
            expect(component.instance().props.message).toBe(expectedCustomProps.message);
            expect(component.instance().props.redirect).toBe(expectedCustomProps.redirect);
            expect(component.instance().props.redirectPath).toBe(expectedDefaultProps.redirectPath);
        });

        it('Don\'t render redirect link', () => {
            expect(component.find(Link)).toHaveLength(0);
        });

        it('Matching snapshot' , () => {
            const tree = toJson(component); 
            
            expect(tree).toMatchSnapshot();
        });
    });
});
