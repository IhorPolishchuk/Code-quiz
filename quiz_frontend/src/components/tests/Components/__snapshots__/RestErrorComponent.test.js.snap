// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestErrorComponent RestErrorComponent with custom props Matching snapshot 1`] = `
<div
  className="nmc__wrapper"
>
  <div
    className="nmc__content"
  >
    <h2
      className="nmc__error_code"
    >
      500
    </h2>
    <div>
      <div
        className="nmc__text"
      >
        <div>
          server error
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RestErrorComponent RestErrorComponent with default props Matching snapshot 1`] = `
<div
  className="nmc__wrapper"
>
  <div
    className="nmc__content"
  >
    <h2
      className="nmc__error_code"
    >
      404
    </h2>
    <div>
      <div
        className="nmc__text"
      >
        <div>
          Page was not found
        </div>
        <div>
          <div>
            You feel lost?
          </div>
          <Link
            className="button__back"
            replace={false}
            to="/"
          >
            Back Home
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;
